File: CarePackage/Software/Overwolf.cs
namespace CarePackage.Software;

public class Overwolf : BaseSoftware
{
    public override string           Key            { get; set; } = "overwolf";
    public override string           Name           { get; set; } = "Overwolf";
    public override SoftwareCategory Category       { get; set; } = SoftwareCategory.Gaming;
    public override string           DownloadName   { get; set; } = "OverwolfSetup.zip";
    public override bool             IsArchive      { get; set; } = true;
    public override bool             ShouldCacheUrl { get; set; } = true;
    public override bool             RequiresAdmin  { get; set; } = false;
    public override Bitmap           Icon           { get; set; } = Resources.Icons.overwolf;
    public override string           Homepage       { get; set; } = "https://overwolf.com";

    private readonly HttpClient _http;

    public Overwolf(IServiceProvider services)
    {
        _http = services.GetKeyedService<HttpClient>("MimicBrowser")!;
    }
    
    public override async Task<string> GetDownloadUrlAsync(CancellationToken ct)
    {
        var res = await _http.GetAsync("https://content.overwolf.com/downloads/setup/latest/regular.html", ct);

        res.EnsureSuccessStatusCode();

        var downloadUrlPattern = new Regex(@"https://setup-overwolf-com\.akamaized\.net/\d{1,}\.\d{1,}\.\d{1,}\.\d{1,}/OverwolfSetup\.zip");
        var html               = await res.Content.ReadAsStringAsync(ct);
        var match              = downloadUrlPattern.Match(html);
        
        DownloadUrlResolveException.ThrowUnless(match.Success);
        
        return match.Groups[0].Value;
    }
}

File: packages/app/src/services/software/definitions/overwolf.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'overwolf';
	public name = 'Overwolf';
	public category = [SoftwareCategory.Gaming];
	public downloadName = 'OverwolfSetup.zip';
	public isArchive = true;
	public shouldCacheUrl = true;
	public icon = 'overwolf.png';
	public homepage = 'https://overwolf.com';

	public async resolveDownloadUrl() {
		const res = await fetch('https://content.overwolf.com/downloads/setup/latest/regular.html');
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/https:\/\/setup-overwolf-com\.akamaized\.net\/\d+\.\d+\.\d+\.\d+\/OverwolfSetup\.zip/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(match[0]);
	}
}
